src/relax/transform/remove_unused_outputs.cc

```diff
@@ -0,0 +1,326 @@
+                           old_func_sinfo->purity);
+
+  auto write_ptr = func.CopyOnWrite();
+  write_ptr->struct_info_ = new_sinfo;
+  write_ptr->body = new_body;
+
+  return func;
+}
+
+class CallSiteMutator : public ExprMutator {
+ public:
+  CallSiteMutator(PMap<GlobalVar, std::function<Expr(Call)>> callsite_updaters)
+      : callsite_updaters_(callsite_updaters) {}
+
+  using ExprMutator::VisitExpr_;
+
+  Expr VisitExpr_(const CallNode* op) override {
+    auto node = Downcast<Call>(ExprMutator::VisitExpr_(op));
+
+    if (auto gvar = node->op.as<GlobalVar>()) {
+      if (auto it = callsite_updaters_.find(gvar.value()); it != callsite_updaters_.end()) {
+        return it->second(node);
+      }
+    }
+
+    return node;
+  }
+
+  PMap<GlobalVar, std::function<Expr(Call)>> callsite_updaters_;
+};
+
+}  // namespace
+
+namespace transform {
+
+Pass RemoveUnusedOutputs() {
+  runtime::TypedPackedFunc<IRModule(IRModule, PassContext)> pass_func =
+      [=](IRModule mod, PassContext pc) -> IRModule {
+    auto usage = PartialTupleUsageCollector::Collect(mod);
+
+    if (usage.empty()) {
+      // Early bail-out if there are no updates to make.
+      return mod;
+    }
+
+    PMap<GlobalVar, std::function<Expr(Call)>> callsite_updaters;
+
+    {
+      IRModule new_callees;
+
+      for (const auto& [gvar, base_func] : mod->functions) {
+        if (auto func = base_func.as<Function>()) {
+          if (auto it = usage.find(gvar); it != usage.end()) {
+            const auto& usage_mask = it->second;
+            auto new_func = UpdateCallee(func.value(), usage_mask);
+
+            GlobalVar new_gvar(gvar->name_hint, new_func->checked_type_);
+            new_gvar->struct_info_ = new_func->struct_info_;
+            new_callees->Add(new_gvar, new_func);
+
+            callsite_updaters[gvar] = [old_gvar = gvar, new_gvar, usage_mask](Call call) -> Expr {
+              ICHECK(call->op.same_as(old_gvar)) << "InternalError: "
+                                                 << "Updater should be applied to " << old_gvar
+                                                 << ", but was applied to " << call->op;
+
+              auto old_call_sinfo = call->struct_info_.as<TupleStructInfoNode>();
+              ICHECK(old_call_sinfo)
+                  << "InternalError: "
+                  << "Updater should be applied to Call producing an output tuple, "
+                  << "but " << call << " has struct info " << call->struct_info_;
+              CHECK_EQ(usage_mask.size(), old_call_sinfo->fields.size())
+                  << "Function " << call->op << " produces " << usage_mask.size() << " outputs, "
+                  << "but " << call << " was used in a context expecting "
+                  << old_call_sinfo->fields.size() << " outputs.";
+
+              Call new_call(new_gvar, call->args);
+
+              int num_outputs_used = 0;
+              for (bool used : usage_mask) {
+                num_outputs_used += used;
```

Comment:
I'm surprised you don't get a warning for this cast.

Reply:
IIRC, the usual automatic integer-conversion warnings occur from `-Wnarrowing`, which checks if there's a conversion between integer types that throws away information.  Since every value that is allowed in `bool` can be uniquely written in `int`, there is no loss of information.